Fix various compiler warnings

krzysz00 opened the pull request:
- Remove unused variables and class members in conv generator
- Change several if/else if/... blocks to switch/case to remove
- warnings about potentially uninitialized variables and to give us a
- warning if we add a case in the future
- Turn some can't happen cases into llvm_unreachable() so they crash
- the compiler if they do, in fact, happen, instead of plowing ahead
- with an error message
- Fix random utilities by:
  - Using an abs() that won't truncate
  - Performing certain calculations with doubles since RAND_MAX is not
    representable by a float on 64-bit platforms

The remaining pile of warnings that's our responsibility is in
rocm-runtime-wrappers.cpp , which has extern "C" functions that export
templated types that probably shouldn't be extern "C" (the non 1D
StridedMemRefType<>s) - it's not clear what to do about those.

krzysz00 committed:
Fix various compiler warnings

- Remove unused variables and class members in conv generator
- Change several if/else if/... blocks to switch/case to remove
- warnings about potentially uninitialized variables and to give us a
- warning if we add a case in the future
- Turn some can't happen cases into llvm_unreachable() so they crash
- the compiler if they do, in fact, happen, instead of plowing ahead
- with an error message
- Fix random utilities by:
  - Using an abs() that won't truncate
  - Performing certain calculations with doubles since RAND_MAX is not
    re presentable by a float on 64-bit platforms

The remaining pile of warnings that's our responsibility is in
rocm-runtime-wrappers.cpp , which has extern "C" functions that export
templated types that probably shouldn't be extern "C" (the non 1D
StridedMemRefType<>s) - it's not clear what to do about those.

whchung commented:
@krzysz00 please help specify the compiler version you experience warnings.

whchung commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -5135,17 +5153,21 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<miopen::GridwiseGemm
           GemmABlockCopyThreadSliceLengths_GemmM;
+      break;
+    case GemmG:
+      llvm_unreachable("Group matrix unimplemented");
```
The message should be something like : " vector loads / stores are not possible on GemmG dimension ".

whchung commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -5169,16 +5191,21 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<miopen::GridwiseGemm
           GemmBBlockCopyThreadSliceLengths_GemmN;
+      break;
+    case GemmG:
+      llvm_unreachable("Group matrix unimplemented");
```
The message should be something like : " vector loads / stores are not possible on GemmG dimension ".

whchung commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -6211,17 +6242,21 @@ struct GridwiseGemmV2RewritePattern
           GemmBBlockCopyThreadSliceLengths_GemmN;
+      break;
+    case GemmG:
+      llvm_unreachable("Group matrix not supported");
```
The message should be something like : " vector loads / stores are not possible on GemmG dimension ".

krzysz00 replied:
Thanks - it was hard to get context for what that case meant.

whchung commented on mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp:
```diff
@@ -36,6 +36,11 @@ Conv2dGenerator::Conv2dGenerator(
              inputLayout,
              outputLayout,
              kernelName,
+             -1,
```
@jerryyin please help review this file.

whchung commented:
For those non-1D `StridedMemRef` `extern "C"` functions, they are hacks put together back in late 2019, due to some very unfortunate design decisions inside HIP memory allocation APIs IIRC. We'll need to revisit them.

+ @deven-amd for awareness.

whchung commented on mlir/lib/ExecutionEngine/ROCm/rocm-runtime-wrappers.cpp:
```diff
@@ -390,7 +390,8 @@ short randomIntegerValue(short min, short max) {
   if (min == max)
     return (float)min;
-  return (max - min) * (float)(std::rand()) / RAND_MAX + (float)min;
+  return (float)((max - min) * (double)(std::rand()) / (double)RAND_MAX) +
```
how about use `static_cast`?

krzysz00 replied:
... good point, will fix

krzysz00 commented:
> 
> 
> @krzysz00 please help specify the compiler version you experience warnings.
The compiler version is clang 12, as shipped with rocm, and also clang 10, as installed in our Dockerfile

krzysz00 committed:
Merge branch 'miopen-dialect' into fix-misc-warnings

krzysz00 committed:
Address review comments

- Make the unreachable() message more accurate
- Use static_cast in the rand() function

jerryyin commented on mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp:
```diff
@@ -36,6 +36,11 @@ Conv2dGenerator::Conv2dGenerator(
+             {},
+             {},
+             {},
+             -1,
              -1} {}
```
@krzysz00 Your changes LGTM

@ltqin The refactor from @krzysz00 caught my attention about the last two member variables: `filterHeight` and `filterWidth`, first introduced [here](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/pull/246/files). I don't think it is the right way to store this duplicated information as those should be available from `filterDimension` vector. Let me know if you think differently on this, otherwise please create a PR/ticket and get rid of those two member variables.

jerryyin replied:
Hearing no reply, recorded this in a ticket https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/255

jerryyin commented on mlir/include/mlir/Dialect/MIOpen/XdlopsCodeSelection.h:
```diff
@@ -658,7 +664,7 @@ struct XdlopsCodeSelection {
       k_base = 2;
     } else {
-      llvm::errs() << "Unsupported case as mfmaInstr not selected!\n";
+      llvm_unreachable("Unsupported case as mfmaInstr not selected!\n");
```
Will this crash the program, and would you suggest using this to intentionally crash ourselves? I figure that our project overall doesn't yet have a consistent way to handle errors.
